### ISSUE TITLE
Graphql Validation Error

### ISSUE THREAD

**Issue by pyros2097** [graphql-go/graphql#104]
I'm getting this weird error
```go
Fragment "F3" cannot be spread here as objects of type "TasksSetPayload" can never be of type "TasksSetPayload".

Variable "$input_0" of type "UserLoginInput!" used in position expecting type "UserLoginInput!".
```
But the same query works on graphiql but fails in react-relay.
Maybe the validation isn't working properly because If i disable validation it seems to work.
Need to dig more into this.

**Comment by pyros2097** [comment]
This was fixed is my commits 43507b96283fa8985a0d7eae5682082cb387543f and 71c79f28a74337c11d0e6b3ab86e7ed9cb3df532.

This seems to happen mainly because we are running graphql in a http server request and this causes the race condition #84 and because of this when you the validate in rules.go you check `varType == expected` Type which wont be equal because they are not the same objects so for now making it work I've added this `varType.Name() == expectedType.Name()`

We need to find a way to use concurrent maps in graphql go to prevent accidental race conditions and that should solve most of these issues

**Comment by sogko** [comment]
Thanks for the PR @pyros2097 👍🏻
